libnetwork/drivers/bridge/port_mapping_linux.go

```diff
@@ -447,31 +500,177 @@ func attemptBindHostPorts(
 	}
 
 	res := make([]portBinding, 0, len(cfg))
+	defer func() {
+		if retErr != nil {
+			for _, pb := range res {
+				if pb.boundSocket != nil {
+					if err := pb.boundSocket.Close(); err != nil {
+						log.G(context.TODO()).Warnf("Failed to close port binding for %s: %s", pb, err)
+					}
+				}
+				// TODO(robmry) - this is only needed because the userland proxy may have
+				//  been started for SCTP. If a bound socket is passed to the proxy after
+				//  iptables rules have been configured (as it is for TCP/UDP), remove this.
+				if pb.stopProxy != nil {
+					if err := pb.stopProxy(); err != nil {
+						log.G(context.TODO()).Warnf("Failed to stop proxy for %s: %s", pb, err)
+					}
+				}
+			}
+		}
+	}()
+
 	for _, c := range cfg {
-		pb := portBinding{PortBinding: c.GetCopy()}
+		var pb portBinding
 		if c.disableNAT {
+			pb = portBinding{PortBinding: c.GetCopy()}
 			pb.HostPort = 0
+			pb.HostPortEnd = 0
 		} else {
-			pb.stopProxy, err = startProxy(c.Proto.String(), c.childHostIP, port, c.IP, int(c.Port), proxyPath)
+			switch proto {
+			case "tcp":
+				pb, err = bindTCPOrUDP(c, port, syscall.SOCK_STREAM, syscall.IPPROTO_TCP)
+			case "udp":
+				pb, err = bindTCPOrUDP(c, port, syscall.SOCK_DGRAM, syscall.IPPROTO_UDP)
+			case "sctp":
+				if proxyPath == "" {
+					pb, err = bindSCTP(c, port)
+				} else {
+					// TODO(robmry) - it's not currently possible to pass a bound SCTP port
+					//  to the userland proxy, because the proxy is not able to convert the
+					//  file descriptor into an sctp.SCTPListener (fd is an unexported member
```

Comment:
Would changes to the sctp library we're using mean that we could get rid of this special case? I also noticed there's a `ListenSCTPext` variant that allows setting some options (maybe that one could be expanded?); https://github.com/ishidawataru/sctp/blob/7ff4192f6ff2b4b181ec7dce269bdf4076ff90a0/sctp_linux.go#L174

ISTR @ishidawataru is a colleague of @AkihiroSuda (but regardless, it's an open source project, so we can contribute / propose changes if that would help us)